packages/common/src/internal-workflow/enums-helpers.ts

```diff
@@ -0,0 +1,294 @@
+  ShortStringEnumKey extends RemovePrefix<Prefix, ProtoEnumKey> = Exclude<
+    RemovePrefix<Prefix, ProtoEnumKey>,
+    Unspecified
+  >,
+  StringEnumType extends ProtoConstObjectOfStringsEnum<
+    ShortStringEnumKey,
+    Prefix,
+    Unspecified
+  > = ProtoConstObjectOfStringsEnum<ShortStringEnumKey, Prefix, Unspecified>,
+  MapTable extends ProtoEnumToConstObjectOfStringMapTable<
+    StringEnumType,
+    ProtoEnumValue,
+    ProtoEnum,
+    ProtoEnumKey,
+    Prefix,
+    Unspecified,
+    ShortStringEnumKey
+  > = ProtoEnumToConstObjectOfStringMapTable<
+    StringEnumType,
+    ProtoEnumValue,
+    ProtoEnum,
+    ProtoEnumKey,
+    Prefix,
+    Unspecified,
+    ShortStringEnumKey
+  >,
+>(
+  mapTable: MapTable,
+  prefix: Prefix
+): [
+  (
+    input: ShortStringEnumKey | `${Prefix}${ShortStringEnumKey}` | ProtoEnumValue | null | undefined
+  ) => ProtoEnumValue | undefined, //
+  (input: ProtoEnumValue | null | undefined) => ShortStringEnumKey | undefined, //
+] {
+  const reverseTable: Record<ProtoEnumValue, ShortStringEnumKey> = Object.fromEntries(
+    Object.entries(mapTable).map(([k, v]) => [v, k])
+  );
+  const hasUnspecified = (mapTable as any)['UNSPECIFIED'] === 0 || (mapTable as any)[`${prefix}UNSPECIFIED`] === 0;
+
+  function isShortStringEnumKeys(x: unknown): x is ShortStringEnumKey {
+    return typeof x === 'string' && x in mapTable;
+  }
+
+  function isNumericEnumValue(x: unknown): x is ProtoEnum[keyof ProtoEnum] {
+    return typeof x === 'number' && x in reverseTable;
+  }
+
+  function encode(
+    input: ShortStringEnumKey | `${Prefix}${ShortStringEnumKey}` | ProtoEnumValue | null | undefined
+  ): ProtoEnumValue | undefined {
+    if (input == null) {
+      return undefined;
+    } else if (typeof input === 'string') {
+      let shorten: string = input;
+      if (shorten.startsWith(prefix)) {
+        shorten = shorten.slice(prefix.length);
+      }
+      if (isShortStringEnumKeys(shorten)) {
+        return mapTable[shorten];
+      }
+      throw new TypeError(`Invalid enum value: '${input}'`);
+    } else if (typeof input === 'number') {
+      return input;
+    } else {
+      throw new TypeError(`Invalid enum value: '${input}' of type ${typeof input}`);
+    }
+  }
+
+  function decode(input: ProtoEnumValue | null | undefined): ShortStringEnumKey | undefined {
+    if (input == null) {
+      return undefined;
+    } else if (typeof input === 'number') {
+      if (hasUnspecified && input === 0) {
+        return undefined;
+      }
+
+      if (isNumericEnumValue(input)) {
+        return reverseTable[input];
+      }
+
+      // FIXME: What should we do with unknown values?
+      // Throwing may end up causing more problems than it would solve if the decoded values
+      // would actually never get read anwyay. An alternative would be to return a string value
+      // containing the numeric value, something like "unknown_23"...
```

Comment:
I think returning the string with the number in it makes sense to me since it mirrors the behavior you'd get with the pure numerical approach.